csi.proto

```diff
@@ -634,6 +843,32 @@ message NodeGetIdResponse {
   // CO in subsequent `ControllerPublishVolume`.
   // This is a REQUIRED field.
   string node_id = 1;
+
+  // The topological segment(s) this node belongs to. A topological
+  // segment is a sub-division of a cluster, like region, zone, rack,
+  // etc.
+  // Each key must be 63 characters or less and consist of alphanumeric
+  // characters, '-', '_' or '.'.
+  // Keys MUST be case-insensitive. Meaning the keys "Zone" and "zone"
+  // MUST not both exist.
+  // Each value MUST be 63 characters or less and begin and end with an
+  // alphanumeric character with '-', '_', '.', or alphanumerics in
+  // between.
+  // COs may use this information along with the topology information
+  // returned in CreateVolumeResponse to ensure that a given volume is
+  // accessible from a given node when scheduling workloads.
+  // This field is OPTIONAL. If it is not specified, the CO MAY assume
+  // the node is not subject to any topological constraint and MAY
+  // schedule workloads that reference any volume V, such that there are
+  // no topological constraints declared for V.
+  //
+  // Example 1:
+  // Given a volume that is accessible from a single zone, and
+  // accessible_from =
+  //   {"region":{"us-west1"}, "zone":{"us-west1-b"}}
+  // Indicates the node exists within the "region" "us-west1" and the
+  // "zone" "us-west1-b".
+  map<string, string> accessible_from = 2;
```

Comment:
Its possible the SP can only make the volume available on specific nodes in a cluster.

How would an SP figure what nodes are in the cluster and then what adapters are available on those to determine if the SP can expose volumes on those nodes? 

Topology should include host adapter level access for an SP to be able to make a volume available to a host where the container is going to run. Nodes can be provisioned/de-provisioned in the cluster.

Its possible that an SP may have multiple paths to a volume but they be active-active or active-passive. A volume may be created and the SP may return the active path (region/zone/site/rack) for the volume (in the create volume response). Later if the passive path becomes active, how does the SP notify the change.

Reply:
> Its possible the SP can only make the volume available on specific nodes in a cluster.

Yes, the SP defines it's own topology it can choose to segment nodes as it chooses.

> How would an SP figure what nodes are in the cluster and then what adapters are available on those to determine if the SP can expose volumes on those nodes?

SP's should have an internal mechanism to discover and identify nodes. CO's use the `NodeGetId` call to map SP node information to CO node information. The SP defines topology so it is up to the SP to specify which topological segment(s) a node falls in to. The CO doesn't tell the SP what the topology is. Volume availability is an inherent trait of an SP and therefore the SP shares its availability topology information with the CO so that the CO can make appropriate scheduling decisions.

> Topology should include host adapter level access for an SP to be able to make a volume available to a host where the container is going to run. Nodes can be provisioned/de-provisioned in the cluster.

This doesn't make sense, could you clarify what you mean.

> Its possible that an SP may have multiple paths to a volume but they be active-active or active-passive. A volume may be created and the SP may return the active path (region/zone/site/rack) for the volume (in the create volume response). Later if the passive path becomes active, how does the SP notify the change.

If it is manifested as limited availability within the CO's cluster, then the design here should suffice. If the volume is equally accessible from the CO's cluster, then that will be addressed by a follow up proposal which will address spreading of volumes across SP specific failure domains.